lib/puma/signal.rb

```diff
@@ -0,0 +1,51 @@
+module Puma
+  module Signal
+    module_function
+
+    def prepend_handler(sig, &handler)
+      name = signame(sig)
+      custom_signal_handlers[signame] ||= []
+      custom_signal_handlers[signame].prepend(handler)
+    end
+
+    # Signal.trap that does not replace
+    # the existing puma handlers
+    def trap(sig, handler_str=nil, &handler)
+      name = signame(sig)
+      ::Signal.trap(name) do
+        invoke_custom_signal_handlers(name)
+
+        if handler.respond_to?(:call)
+          yield handler
+        else
+          # dirty hack to call string handlers
```

Comment:
Well, I'm stuck. First of all, I don't like this dirty hack. Besides that, it goes into an infinity loop on the `SIGCHLD`, I don't know exactly why.

@MSP-Greg could I use your assistance?

Reply:
@alexeevit I'll have a look a bit later.  Interesting failures, may have nothing to do with this PR.  Not.sure.

Could you rebase?

Reply:
Done 👌

Reply:
Thanks.  Same problem as before...  Sorry.

Reply:
What do you mean? You thought the rebase could help?

Reply:
Wasn't sure, but I don't recall seeing a PR that 'failed' so consistently with a set of Ruby versions.

Reply:
I've just tested this hack independently of puma and it works fine on ruby 2.5 and fails on all next versions

Reply:
```
def custom_trap(sig, handler_str=nil, &handler)
  new_handler = proc do
    if handler.respond_to?(:call)
      yield handler
    else
      # dirty hack to call string handlers
      current_handler = Signal.trap(sig, handler_str)
      Process.kill(sig, Process.pid)
      Signal.trap(sig, new_handler) # or current_handler, same result
    end
  end

  Signal.trap(sig, new_handler)
end

custom_trap 'CHLD', 'IGNORE'
```

In ruby 2.5 it just exits and in ruby 2.6 it goes into an infinite loop.

Reply:
Anyway, I'd like to have a better way to invoke the string handlers for signals (e.g. DEFAULT, SYSTEM_DEFAULT, IGNORE, ...). But this behavior looks quite interesting.

Reply:
We might have the explanation for the problems seen here with Ruby 2.6+, see https://github.com/puma/puma/issues/3313#issuecomment-1887925130

Reply:
I tested the script in https://github.com/puma/puma/pull/2767#discussion_r873130892 with Ruby 3.3, and it exited right away. This seems to suggest that the `SIGCHLD` handler introduced in Ruby 2.6 via https://github.com/ruby/ruby/commit/054a412d540e7ed2de63d68da753f585ea6616c3 may indeed be the issue.

Note that I tested the script with the `ruby_3_2` branch, and it still got stuck. There is a different bug than the one described in https://github.com/puma/puma/issues/3313, which is actually a Ruby bug already fixed in the `ruby_3_2` branch: https://bugs.ruby-lang.org/issues/19837

Reply:
@stanhu thanks for the info!
However, it still isn't clear, what's the expected fix here. Should we support the feature only for the ruby versions that have the bug fixed? Didn't dig into the nature of the bug, so not sure if there is a workaround for other ruby versions.